crypto/fips/self_test.c

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <openssl/ossl_typ.h>
+#include <openssl/evp.h>
+#include <openssl/ec.h>
+#include <openssl/rsa.h>
+#include <openssl/dsa.h>
+#include <openssl/hmac.h>
+#include "self_test.h"
+
+#define FIPS_STATE_INIT     0
+#define FIPS_STATE_RUNNING  1
+#define FIPS_STATE_SELFTEST 2
+#define FIPS_STATE_ERROR    3
+
+static int FIPS_state = FIPS_STATE_INIT;
+static unsigned char fixed_key[32] = {0};
+static SELF_TEST_CB SELF_TEST_cb = NULL;
+
+void SELF_TEST_EVENT_init(ST_EVENT *ev)
+{
+    size_t n = 0;
+    OSSL_PARAM param_end = OSSL_PARAM_END;
+
+    if (ev == NULL)
+        return;
+
+    if (SELF_TEST_cb != NULL) {
+        ev->params[n++] = OSSL_PARAM_construct_int("phase", &ev->phase, NULL);
+        ev->params[n++] = OSSL_PARAM_construct_int("type",  &ev->type, NULL);
+        ev->params[n++] = OSSL_PARAM_construct_int("desc",  &ev->desc, NULL);
+    }
+    ev->params[n++] = param_end;
```

Comment:
Can't this just be `ev->params[n++] = OSSL_PARAM_END`

Reply:
No :(

OSSL_PARAM_END is defined as { ... }.  It is worth considering making it a global const variable so such an assignment does work.

Reply:
#8704.  It was time.  We've talking about making such a constructor so many times that I'm surprised it took this long to actually make the PR...